File: src/YgoPicProxy.hs
{-# LANGUAGE OverloadedStrings #-}

module YgoPicProxy
  ( AppState
  , newAppState
  , app
  , worker
  , parseId
  ) where

import Control.Concurrent.Chan (Chan, newChan, readChan, writeChan)
import Control.Exception (IOException, SomeException, bracket, bracketOnError, catch, finally, try)
import Control.Monad (forever)
import qualified Data.ByteString.Lazy as BL
import Data.Char (isDigit)
import Data.Text (Text)
import qualified Data.Text as T
import Data.Time.Clock.POSIX (getPOSIXTime)
import Database.SQLite.Simple (Connection, Only (..), execute, execute_, open, query)
import Network.HTTP.Client (Manager, httpLbs, parseRequest, responseBody, responseStatus, responseTimeout, responseTimeoutMicro)
import Network.HTTP.Client.TLS (newTlsManager)
import Network.HTTP.Types (status200, status404, status500)
import Network.Wai (Application, Response, ResponseReceived, pathInfo, responseFile, responseLBS)
import System.Directory (copyFile, createDirectoryIfMissing, doesFileExist, getTemporaryDirectory, removeFile)
import System.FilePath ((</>))
import System.IO (hClose, openTempFile)
import System.Process (callProcess)
import Text.Read (readMaybe)

data AppState = AppState{
  asChan    :: Chan (String, FilePath),
  asDb      :: Connection,
  asManager :: Manager
}

newAppState :: IO AppState
newAppState = do
  createDirectoryIfMissing True "cache"
  conn <- open "ygo-pic-proxy.db"
  initDb conn
  mgr <- newTlsManager
  chan <- newChan
  pure (AppState chan conn mgr)

initDb :: Connection -> IO ()
initDb conn = do
  execute_ conn "PRAGMA journal_mode=WAL;"
  execute_ conn "PRAGMA busy_timeout=5000;"
  execute_ conn "CREATE TABLE IF NOT EXISTS notexist (id TEXT PRIMARY KEY, timestamp TEXT NOT NULL)"

app :: AppState -> Application
app st req respond =
  case pathInfo req of
    [name] ->
      case parseId name of
        Just i  -> handleImage st i respond
        Nothing -> respond (responseLBS status404 [] "not found")
    _ -> respond (responseLBS status404 [] "not found")

handleImage :: AppState -> String -> (Response -> IO ResponseReceived) -> IO ResponseReceived
handleImage st cid respond = do
  let cacheFile = "cache" </> (cid ++ ".jpg")
  cached <- doesFileExist cacheFile
  if cached
    then respond (responseFile status200 [("Content-Type", "image/jpeg")] cacheFile Nothing)
    else fetchMissing
  where
    fetchMissing :: IO ResponseReceived
    fetchMissing = do
      now <- floor <$> getPOSIXTime
      mTs <- getNotExist (asDb st) cid
      case mTs of
        Just ts | now - ts < 3600 -> respond (responseLBS status404 [] "not found")
        _ -> withTempFile ("ygo-" ++ cid ++ ".webp") $ \webpFile -> do
          dl <- try (downloadWebp (asManager st) cid webpFile) :: IO (Either SomeException DownloadResult)
          case dl of
            Left _ -> respond (responseLBS status500 [] "internal error, download failed")
            Right DownloadNotFound -> do
              setNotExist (asDb st) cid now
              respond (responseLBS status404 [] "not found")
            Right DownloadHttpError -> respond (responseLBS status500 [] "internal error, download http error")
            Right DownloadOk ->
              withTempFileOnError ("ygo-" ++ cid ++ ".jpg") $ \jpgFile -> do
                callProcess "magick" [webpFile, jpgFile]
                img <- BL.readFile jpgFile
                writeChan (asChan st) (cid, jpgFile)
                respond (responseLBS status200 [("Content-Type", "image/jpeg")] img)
              `catch` \(_ :: SomeException) ->
                respond (responseLBS status500 [] "internal error, magick exception")

data DownloadResult = DownloadOk | DownloadNotFound | DownloadHttpError

ignoreIO :: IOException -> IO ()
ignoreIO _ = pure ()

removeQuietly :: FilePath -> IO ()
removeQuietly p = catch (removeFile p)  ignoreIO

acquireTempFile :: String -> IO FilePath
acquireTempFile template = do
  tmpDir <- getTemporaryDirectory
  (p, h) <- openTempFile tmpDir template
  hClose h
  pure p

withTempFile :: String -> (FilePath -> IO a) -> IO a
withTempFile template = bracket (acquireTempFile template) removeQuietly

withTempFileOnError :: String -> (FilePath -> IO a) -> IO a
withTempFileOnError template = bracketOnError (acquireTempFile template) removeQuietly

downloadWebp :: Manager -> String -> FilePath -> IO DownloadResult
downloadWebp mgr cid dest = do
  req0 <- parseRequest ("https://cdn.233.momobako.com/ygoimg/ygopro/" ++ cid ++ ".webp!/format/webp/fw/400/quality/85")
  let req = req0 { responseTimeout = responseTimeoutMicro 30000000 }
  resp <- httpLbs req mgr
  let st = responseStatus resp
  if st == status404 then
    pure DownloadNotFound
  else if st /= status200 then
    pure DownloadHttpError
  else do
    BL.writeFile dest (responseBody resp)
    pure DownloadOk

getNotExist :: Connection -> String -> IO (Maybe Integer)
getNotExist conn cid = do
  rows <- query conn "SELECT timestamp FROM notexist WHERE id = ?" (Only cid) :: IO [Only String]
  case rows of
    Only tsStr : _ -> pure (readMaybe tsStr)
    []             -> pure Nothing

setNotExist :: Connection -> String -> Integer -> IO ()
setNotExist conn cid ts =
  execute conn "INSERT OR REPLACE INTO notexist (id, timestamp) VALUES (?, ?)" (cid, show ts)

worker :: AppState -> IO ()
worker st = forever $ do
  (cid, tmpJpg) <- readChan (asChan st)
  flip finally (removeQuietly tmpJpg) $ do
    let cacheJpg = "cache" </> (cid ++ ".jpg")
    exists <- doesFileExist cacheJpg
    if exists then pure ()
    else catch (copyFile tmpJpg cacheJpg) ignoreIO

parseId :: Text -> Maybe String
parseId name = do
  s <- T.stripSuffix ".jpg" name
  let str = T.unpack s
  if not (null str) && length str <= 10 && all isDigit str then
    Just str
  else Nothing

File: src/ygoPicProxy.js
import express from 'express';
import fs from 'node:fs/promises';
import fsSync from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import crypto from 'node:crypto';
import { execFile } from 'node:child_process';
import { promisify } from 'node:util';
import { openDb } from './data_access.js';

const execFileAsync = promisify(execFile);

const UPSTREAM_PREFIX = 'https://cdn.233.momobako.com/ygoimg/ygopro/';
const UPSTREAM_SUFFIX = '.webp!/format/webp/fw/400/quality/85';
const DOWNLOAD_TIMEOUT_MS = 30000;
const NOT_EXIST_TTL_SECONDS = 3600;

// ---- id parsing ----

function parseId(name) {
  if (!name.endsWith('.jpg')) return null;
  const s = name.slice(0, -'.jpg'.length);
  if (s.length > 0 && s.length <= 10 && /^[0-9]+$/.test(s)) return s;
  return null;
}

// ---- simple FIFO async queue (Chan equivalent) ----

function AsyncQueue() {
  let self = {
    items: [],
    waiters: [],
    push: function (item) {
      if (self.waiters.length > 0) {
        const resolve = self.waiters.shift();
        resolve(item);
      } else {
        self.items.push(item);
      }
    },
    pop: function () {
      if (self.items.length > 0) {
        return Promise.resolve(self.items.shift());
      }
      return new Promise((resolve) => self.waiters.push(resolve));
    },
  };
  return self;
}

// ---- app state ----

async function createAppState(opts = {}) {
  const cacheDir = opts.cacheDir ?? 'cache';
  const dbPath = opts.dbPath ?? 'ygo-pic-proxy.db';

  await fs.mkdir(cacheDir, { recursive: true });

  const db = openDb(dbPath);

  return {
    db,
    cacheDir,
    queue: AsyncQueue(),
  };
}

// ---- temp file helpers ----

function tempFilePath(prefix, ext) {
  const rand = crypto.randomBytes(8).toString('hex');
  return path.join(os.tmpdir(), `${prefix}-${rand}${ext}`);
}

async function removeQuietly(p) {
  try {
    await fs.unlink(p);
  } catch {
    // ignore, matches Haskell's removeQuietly (ignores IOException)
  }
}

async function fileExists(p) {
  try {
    await fs.access(p, fsSync.constants.F_OK);
    return true;
  } catch {
    return false;
  }
}

// ---- upstream download ----

async function downloadWebp(cid, dest) {
  const url = `${UPSTREAM_PREFIX}${cid}${UPSTREAM_SUFFIX}`;
  const resp = await fetch(url, { signal: AbortSignal.timeout(DOWNLOAD_TIMEOUT_MS) });
  if (resp.status === 404) return 'not-found';
  if (resp.status !== 200) return 'http-error';
  const buf = Buffer.from(await resp.arrayBuffer());
  await fs.writeFile(dest, buf);
  return 'ok';
}

async function runMagick(src, dest) {
  await execFileAsync('magick', [src, dest]);
}

// ---- HTTP responses ----

function notFound(res) {
  res.status(404);
  res.set('Content-Length', String(Buffer.byteLength('not found')));
  res.end('not found');
}

function serverError(res, message) {
  res.status(500);
  res.set('Content-Length', String(Buffer.byteLength(message)));
  res.end(message);
}

async function sendCachedFile(res, cacheFile) {
  const stat = await fs.stat(cacheFile);
  res.status(200);
  res.set('Content-Type', 'image/jpeg');
  res.set('Content-Length', String(stat.size));
  await new Promise((resolve, reject) => {
    const stream = fsSync.createReadStream(cacheFile);
    stream.on('error', reject);
    stream.on('end', resolve);
    stream.pipe(res);
  });
}

function sendImage(res, buffer) {
  res.status(200);
  res.set('Content-Type', 'image/jpeg');
  res.set('Content-Length', String(buffer.length));
  res.end(buffer);
}

// ---- core request handling ----

async function handleImage(state, cid, res) {
  const cacheFile = path.join(state.cacheDir, `${cid}.jpg`);

  if (await fileExists(cacheFile)) {
    await sendCachedFile(res, cacheFile);
    return;
  }

  const now = Math.floor(Date.now() / 1000);
  const ts = state.db.getNotExist(cid);
  if (ts !== null && now - ts < NOT_EXIST_TTL_SECONDS) {
    notFound(res);
    return;
  }

  const webpFile = tempFilePath(`ygo-${cid}`, '.webp');
  try {
    let result;
    try {
      result = await downloadWebp(cid, webpFile);
    } catch {
      serverError(res, 'internal error, download failed');
      return;
    }

    if (result === 'not-found') {
      state.db.setNotExist(cid, now);
      notFound(res);
      return;
    }
    if (result === 'http-error') {
      serverError(res, 'internal error, download http error');
      return;
    }

    const jpgFile = tempFilePath(`ygo-${cid}`, '.jpg');
    try {
      await runMagick(webpFile, jpgFile);
      const img = await fs.readFile(jpgFile);
      state.queue.push({ cid, jpgFile });
      sendImage(res, img);
    } catch {
      await removeQuietly(jpgFile);
      serverError(res, 'internal error, magick exception');
    }
  } finally {
    await removeQuietly(webpFile);
  }
}

// ---- express app ----

function createApp(state) {
  const app = express();
  app.disable('x-powered-by');
  app.disable('etag');

  app.use((req, res) => {
    const p = req.path.startsWith('/') ? req.path.slice(1) : req.path;
    if (p.includes('/')) {
      notFound(res);
      return;
    }
    const id = parseId(p);
    if (id === null) {
      notFound(res);
      return;
    }
    handleImage(state, id, res).catch(() => {
      if (!res.headersSent) {
        res.status(500);
      }
      res.end();
    });
  });

  return app;
}

// ---- background worker: moves converted temp files into the disk cache ----

async function worker(state) {
  for (;;) {
    const { cid, jpgFile } = await state.queue.pop();
    try {
      const cacheFile = path.join(state.cacheDir, `${cid}.jpg`);
      if (!(await fileExists(cacheFile))) {
        try {
          await fs.copyFile(jpgFile, cacheFile);
        } catch {
          // ignore, matches Haskell's ignoreIO
        }
      }
    } finally {
      await removeQuietly(jpgFile);
    }
  }
}

export { parseId, createAppState, createApp, worker };
